### ISSUE TITLE
Extending a cluster to an extra node never seems to catch up

### ISSUE THREAD

**Issue by jtorreke** [openvstorage/arakoon#191]
The scenario:

We had a 5 node arakoon cluster and used the framework tools to move one of the members to a different node. The plan was to extend to cluster with a 6th node to afterwards shrink to a 5 node cluster again. We noticed that the new member was never able to catch up with the cluster, resulting in a more or less endless loop. We let the new member try to catch up for +12 hours, but to no avail. It must also be noted that one of the existing members of the cluster was down before we triggered the extend to a 6th node, hence the log file contains connection refused messages for node 172.17.16.8.

Some log file snippets:

```
2017-07-25 05:47:04 802887 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626081 - info - CATCHUP start: I'm @ 22280560 and N0RCG4vXgbP0w8LQ is more recent
2017-07-25 05:47:04 802906 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626082 - info - catchup_tlog 22280560 from N0RCG4vXgbP0w8LQ
2017-07-25 05:47:04 854183 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626083 - info - loop_entries
2017-07-25 05:47:04 946255 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626084 - info - remote_nodestream :: last_seen = Some ("22280579")
2017-07-25 05:47:04 946299 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626085 - info - loop_parts done
2017-07-25 05:47:04 946337 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626086 - info - catchup_tlog completed
2017-07-25 05:47:04 946358 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626087 - info - CATCHUP phase 1 done (until_i = 22280579); now the store
2017-07-25 05:47:04 946372 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626088 - info - replaying log to store
2017-07-25 05:47:04 951188 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626089 - info - Replaying tlog file: 439.tlog [22280281,...] (195/195)
2017-07-25 05:47:05 921475 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626090 - info - establishing connection to (172.17.16.8,26455) (timeout = 5.000000)
2017-07-25 05:47:05 921638 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626091 - info - __open_connection to (172.17.16.8,26455) failed: Unix.Unix_error(Unix.ECONNREFUSED, "connect", "")
2017-07-25 05:47:06 608391 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626092 - info - Completed replay of 439.tlog, took 1.657176 seconds, 0 to go
2017-07-25 05:47:06 650158 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626093 - info - catchup_store completed, store is @ Some ("22280579")
2017-07-25 05:47:06 650182 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626094 - info - CATCHUP end
2017-07-25 05:47:06 650279 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626095 - info - CATCHUP start: I'm @ 22280579 and wZ4GAcmR2MZl5x7S is more recent
2017-07-25 05:47:06 650286 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626096 - info - catchup_tlog 22280579 from wZ4GAcmR2MZl5x7S
2017-07-25 05:47:06 654715 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626097 - info - loop_entries
2017-07-25 05:47:06 806162 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626098 - info - remote_nodestream :: last_seen = Some ("22280586")
2017-07-25 05:47:06 806184 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626099 - info - loop_parts done
2017-07-25 05:47:06 806205 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626100 - info - catchup_tlog completed
2017-07-25 05:47:06 806228 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626101 - info - CATCHUP phase 1 done (until_i = 22280586); now the store
2017-07-25 05:47:06 806235 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626102 - info - replaying log to store
2017-07-25 05:47:06 809906 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626103 - info - Replaying tlog file: 439.tlog [22280281,...] (195/195)
2017-07-25 05:47:07 037526 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626104 - info - Completed replay of 439.tlog, took 0.227597 seconds, 0 to go
2017-07-25 05:47:07 081012 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626105 - info - catchup_store completed, store is @ Some ("22280586")
2017-07-25 05:47:07 081033 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626106 - info - CATCHUP end
2017-07-25 05:47:07 185386 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626107 - info - establishing connection to (172.17.16.8,26455) (timeout = 5.000000)
2017-07-25 05:47:07 185647 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626108 - info - __open_connection to (172.17.16.8,26455) failed: Unix.Unix_error(Unix.ECONNREFUSED, "connect", "")
2017-07-25 05:47:07 227959 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626109 - info - CATCHUP start: I'm @ 22280586 and N0RCG4vXgbP0w8LQ is more recent
2017-07-25 05:47:07 227983 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626110 - info - catchup_tlog 22280586 from N0RCG4vXgbP0w8LQ
2017-07-25 05:47:07 239730 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626111 - info - loop_entries
2017-07-25 05:47:07 296488 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626112 - info - remote_nodestream :: last_seen = Some ("22280587")
2017-07-25 05:47:07 296506 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626113 - info - loop_parts done
2017-07-25 05:47:07 296522 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626114 - info - catchup_tlog completed
2017-07-25 05:47:07 296529 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626115 - info - CATCHUP phase 1 done (until_i = 22280587); now the store
2017-07-25 05:47:07 296551 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626116 - info - replaying log to store
2017-07-25 05:47:07 299951 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626117 - info - Replaying tlog file: 439.tlog [22280281,...] (195/195)
2017-07-25 05:47:07 359810 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626118 - info - Completed replay of 439.tlog, took 0.059834 seconds, 0 to go
2017-07-25 05:47:07 360198 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626119 - info - catchup_store completed, store is @ Some ("22280587")
2017-07-25 05:47:07 360210 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626120 - info - CATCHUP end
2017-07-25 05:47:07 371226 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626121 - info - CATCHUP start: I'm @ 22280587 and wZ4GAcmR2MZl5x7S is more recent
2017-07-25 05:47:07 371253 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626122 - info - catchup_tlog 22280587 from wZ4GAcmR2MZl5x7S
2017-07-25 05:47:07 373095 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626123 - info - loop_entries
2017-07-25 05:47:07 479117 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626124 - info - remote_nodestream :: last_seen = Some ("22280588")
2017-07-25 05:47:07 479141 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626125 - info - loop_parts done
2017-07-25 05:47:07 479159 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626126 - info - catchup_tlog completed
2017-07-25 05:47:07 479168 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626127 - info - CATCHUP phase 1 done (until_i = 22280588); now the store
2017-07-25 05:47:07 479175 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626128 - info - replaying log to store
2017-07-25 05:47:07 483116 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626129 - info - Replaying tlog file: 439.tlog [22280281,...] (195/195)
2017-07-25 05:47:07 539037 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626130 - info - Completed replay of 439.tlog, took 0.055894 seconds, 0 to go
2017-07-25 05:47:07 587151 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626131 - info - catchup_store completed, store is @ Some ("22280588")
2017-07-25 05:47:07 587172 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626132 - info - CATCHUP end
```

To have a more clear view on the amount of updates the new member was running behind:

```
2017-07-25 05:48:13 161195 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626653 - info - CATCHUP start: I'm @ 22280757 and wZ4GAcmR2MZl5x7S is more recent
2017-07-25 05:48:13 169220 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626659 - info - CATCHUP phase 1 done (until_i = 22280759); now the store
2017-07-25 05:48:13 268636 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626664 - info - CATCHUP end
2017-07-25 05:48:13 268955 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626665 - info - CATCHUP start: I'm @ 22280759 and wZ4GAcmR2MZl5x7S is more recent
2017-07-25 05:48:13 285894 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626671 - info - CATCHUP phase 1 done (until_i = 22280761); now the store
2017-07-25 05:48:13 412463 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626676 - info - CATCHUP end
2017-07-25 05:48:13 412893 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626677 - info - CATCHUP start: I'm @ 22280761 and wZ4GAcmR2MZl5x7S is more recent
2017-07-25 05:48:13 435778 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626683 - info - CATCHUP phase 1 done (until_i = 22280762); now the store
2017-07-25 05:48:13 623228 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626690 - info - CATCHUP end
2017-07-25 05:48:13 623528 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626691 - info - CATCHUP start: I'm @ 22280762 and N0RCG4vXgbP0w8LQ is more recent
2017-07-25 05:48:13 736298 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626697 - info - CATCHUP phase 1 done (until_i = 22280764); now the store
2017-07-25 05:48:13 839142 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626702 - info - CATCHUP end
2017-07-25 05:48:13 839181 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626703 - info - CATCHUP start: I'm @ 22280764 and wZ4GAcmR2MZl5x7S is more recent
2017-07-25 05:48:13 847774 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626709 - info - CATCHUP phase 1 done (until_i = 22280766); now the store
2017-07-25 05:48:14 027051 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626714 - info - CATCHUP end
2017-07-25 05:48:25 909635 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626740 - info - CATCHUP start: I'm @ 22280766 and N0RCG4vXgbP0w8LQ is more recent
2017-07-25 05:48:26 053015 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626746 - info - CATCHUP phase 1 done (until_i = 22280829); now the store
2017-07-25 05:48:27 210702 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626753 - info - CATCHUP end
2017-07-25 05:48:27 210738 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626754 - info - CATCHUP start: I'm @ 22280829 and wZ4GAcmR2MZl5x7S is more recent
2017-07-25 05:48:27 227744 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626760 - info - CATCHUP phase 1 done (until_i = 22280837); now the store
2017-07-25 05:48:27 449894 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626765 - info - CATCHUP end
2017-07-25 05:48:27 860400 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626768 - info - CATCHUP start: I'm @ 22280837 and wZ4GAcmR2MZl5x7S is more recent
2017-07-25 05:48:27 891243 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626774 - info - CATCHUP phase 1 done (until_i = 22280838); now the store
2017-07-25 05:48:28 134927 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626779 - info - CATCHUP end
2017-07-25 05:48:39 371910 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626803 - info - CATCHUP start: I'm @ 22280838 and hVEIul3TXTpCAWwg is more recent
2017-07-25 05:48:39 674368 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626809 - info - CATCHUP phase 1 done (until_i = 22280861); now the store
2017-07-25 05:48:41 339351 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626818 - info - CATCHUP end
2017-07-25 05:48:41 339401 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626819 - info - CATCHUP start: I'm @ 22280861 and wZ4GAcmR2MZl5x7S is more recent
2017-07-25 05:48:41 388302 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626825 - info - CATCHUP phase 1 done (until_i = 22280862); now the store
2017-07-25 05:48:42 229435 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626830 - info - CATCHUP end
2017-07-25 05:48:42 229831 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626832 - info - CATCHUP start: I'm @ 22280862 and hVEIul3TXTpCAWwg is more recent
2017-07-25 05:48:42 350230 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626839 - info - CATCHUP phase 1 done (until_i = 22280863); now the store
2017-07-25 05:48:42 473137 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626844 - info - CATCHUP end
2017-07-25 05:48:42 716423 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626845 - info - CATCHUP start: I'm @ 22280863 and wZ4GAcmR2MZl5x7S is more recent
2017-07-25 05:48:42 745541 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626851 - info - CATCHUP phase 1 done (until_i = 22280864); now the store
2017-07-25 05:48:43 230890 -0400 - NY1SRV0012 - 11327/0000 - arakoon - 1626856 - info - CATCHUP end
```

**Comment by toolslive** [comment]
Can you compare your scenario to [this](https://github.com/openvstorage/arakoon/blob/1.9/pylabs/test/server/shaky/test_sso_deployment.py#L32) test and highlight the differences ? (I think you're going to have flash backs while reading the comments)

**Comment by jtorreke** [comment]
I found the culprit in the mean time (but forgot I created a ticket on this). The cluster is extended with an extra node, but for some reason, not all nodes were restarted. So some members of the cluster had no knowledge of the new member. If the current master is such a not-restarted node, it seems to refuse the participation of the new node in the cluster and keeps it lagging behind. A restart of all cluster members fixed this behavior.

TL;DR: please add this user error to the statistics

**Comment by toolslive** [comment]
;)
